images/kindnetd/cmd/kindnetd/subnets.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+)
+
+type subnets struct {
+	pods     sets.String
+	services sets.String
+}
+
+func newSubnets() *subnets {
+	return &subnets{
+		pods:     sets.NewString(),
+		services: sets.NewString(),
+	}
+}
+
+func (s *subnets) All() []string {
+	return s.pods.Union(s.services).List()
+}
+
+func getNoMasqueradeSubnets(clientset *kubernetes.Clientset) []string {
+	s := newSubnets()
+
+	// try getting from kubeadm config
+	kubeadmSubnets := getKubeadmSubnets(clientset)
+	s.pods = s.pods.Union(kubeadmSubnets.pods)
+	s.services = s.services.Union(kubeadmSubnets.services)
```

Comment:
IIRC there is no need to insert the services subnet in the no-masquerade chain,  Tim Hockin  explained here https://github.com/kubernetes/kubernetes/issues/79866#issuecomment-519218155 that pod to service traffic is always masquerade 

All traffic directed to services inside a node is processed by this chain:
`
-A PREROUTING -m comment --comment "kubernetes service portals" -j KUBE-SERVICES 
`
and if it is directed to one service it sets a mark
`-A KUBE-MARK-MASQ -j MARK --set-xmark 0x4000/0x4000`
so this traffic is *always* masquerade
`-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -m mark --mark 0x4000/0x4000 -j MASQUERADE `

I think that this no-masq rule will only affect the traffic from pods to services ips that doesn't exists

Reply:
Sure, this is a carry over from that PR. I can remove looking at services in the morning. I also want to make some improvements to error handling in the route agent.